rfd/0137-unified-resource-view.md

```diff
@@ -0,0 +1,121 @@
+```go
+func (u *unifiedResourceCollector) resourceKind() []types.WatchKind {
+	return []types.WatchKind{
+		{Kind: types.KindNode},
+		{Kind: types.KindDatabaseServer},
+		{Kind: types.KindAppServer},
+		{Kind: types.KindWindowsDesktop},
+	}
+}
+```
+The above example can be expanded to include any Kinds we may need/want to add in our Unified Resource view.
+
+When the watcher inits, we populate the btree with all the available items on startup
+```go
+func (u *uiResourceCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	err := u.getAndUpdateNodes(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = u.getAndUpdateDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	//...
+
+	u.defineCollectorAsInitialized()
+	return nil
+}
+```
+
+If the collector is uninitialized or stale, we will resort to the upstream and send Nodes only.
+
+### ListUnifiedResources grpc endppint
+We will add a new grpc endpoint, `ListUnifiedResources`. It will function pretty closes to how `ListResources` works, except without needing to differentiate between `RequestType`, as every request is a unified resource request. Originally, the thought of just extending `ListResources`, but because of the many new filters that will be added to the request, we don't want to muddy up the `ListResourcesRequest`. Some of the new filters include which types to include (can be multiple), subtypes, etc etc. 
+
+This endpoint will pull resources from our unified resource cache and run through the same filtering/rbac as `ListResources`
+
+### Backward Compatibility
+
+We will leave `ListResources` alone and keep the old individual resource endpoints available until we've vetted the unified resource view. We will remove the individual navigation items from the web ui, but leave the routes/api endpoints until we've gained confidence in the new flow. 
+
+### UX
+The UI for the web will be changing drastically given the fact that all resources will be on the same page and we will be using a Card layout rather than the current Table. However, besides the visual changes, the actual resources listed will still be interacted with the same way. You can still click "Connect" on servers and databases, or Login for apps. The individual functionality for each "Item" will remain the same. Labels will be truncated but viewable with an expand button.
+
+
+
+https://github.com/gravitational/teleport/assets/5201977/9c859a2a-784d-4805-9ba7-9e37a8a4ce77
+
+
+
+Also, filters will be added. Some filters will just be a "pseudo-filter" in the sense that it'll just help populate an advanced search string, (example, clicking a label), but a few new actual filters will be present in the requests. The biggest of them is the `Kinds` array, which will contain a list of kinds that you want to be returned. ["node","app"] will only return nodes and apps. If left empty (the default), all resource kinds will be returned
```

Comment:
I'm having trouble envisioning drop down filter for labels and addresses as well. Is there a design for that? Would it be made searchable? There's just going to be a ton of them so I'm curious to know how it will be implemented (if it will be).

Reply:
I don't think the designs are final (yet?) but you can see the current work here https://www.figma.com/file/JKd0Bgs30AnNfNCLUuzzc2/Unified-Resources?type=design&node-id=0-1&mode=design&t=Z6fdPQrsVJWx6CrU-0

cc: @roraback

Reply:
Update: the dropdowns for labels and addresses will function mostly as an 'input' box that can be used to guide the user toward creating advance search predicate. imagine, an input box on labels for "key" and then it will allow you to type a "value". They are coming after resource pinning so not including in the initial release.